### PR TITLE
Implement KJ Filesystem API on Win32

### PR DESCRIPTION
**Pull request by kentonv**
(WIP warning: I've only tested on MinGW / WINE so far. Gonna let AppVeyor do a real Windows build and see what happens.)

This also officially adds the KJ filesystem API to the non-Ekam build.

This implementation does not attempt to use the hidden Windows NT equivalents to POSIX's openat() family (e.g. [NtCreateFile](https://msdn.microsoft.com/en-us/library/bb432380(v=vs.85).aspx)), because:

1. They are documented as "internal, could change in the future".
2. You have to use LoadLibrary() to obtain the function addresses dynamically, which is weird.
3. It doesn't look like every operation is supported. In particular I could not find a way to implement moves.
4. The APIs are even more hopelessly overcomplicated than regular Win32.

Instead, we keep track of full paths. To prevent the path from becoming inconsistent due to renames, we open directory handles without `FILE_SHARE_DELETE`, meaning any attempt by other processes to move the directory while we still have a handle open will be denied.

### CODE REVIEW

**Comment by kentonv** [comment]
Holy cow, I just found: https://undocumented.ntinternals.net/

Wish I'd seen that earlier. It looks like a "renameat" equivalent *is* in there (using FILE_RENAME_INFORMATION).

Of course, these are all "undocumented, could change" calls...

Not gonna go back and start over now but maybe someday.

**Comment by kentonv** [comment]
After an 18GB VM download and installation of several tools, I was able to run the test in the Visual Studio debugger, at which point it became immediately obvious what the problem was: the same old constexpr bug where the vtable pointer isn't initialized.

Fixed, and now it all works!

@harrishancock Don't feel like you have to review this in intense detail. Especially `filesystem-disk-win32.c++` itself is not that important to our job, since we don't use Windows. If there are bugs in there, we'll deal with them as we find them. (But it already passes the same tests as the Unix implementation.)

**Comment by harrishancock** [comment]
> [...] the same old constexpr bug where the vtable pointer isn't initialized.

Ah, our old friend! I'm glad that's all it turned out to be. :)

Will review today.

File: c++/src/kj/filesystem-disk-generic-test.c++ [hunk 1]
```diff
@@ -19,8 +19,18 @@
...
 // This test compiles filesystem-disk-unix.c++ with various features #undefed, causing it to take
 // different code paths, then runs filesystem-disk-test.c++ against that.
+//
+// This test is only intended to run on Linux, but is intended to make the code behave like it
+// would on a generic flavor of Unix.
+//
+// At present this test only runs under Ekam builds. Integrating it into other builds would be
+// awkward since it #includes filesystem-disk-unix.c++, so it cannot link against that file, but
+// needs to link against the rest of KJ. Ekam "just figures it out", but other build systems would
+// require a lot of work here.
```
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/filesystem-disk-generic-test.c++]: I don't think Ekam builds are part of the capnp release tests currently, but they're pretty thoroughly covered by us in our development environment... :)

File: c++/src/kj/encoding.c++ [hunk 2]
```diff
@@ -247,6 +247,85 @@ EncodingResult<String> decodeUtf32(ArrayPtr<const char32_t> utf16) {
...
 }
 
+namespace {
+
+template <typename To, typename From>
+Array<To> coerceTo(Array<From>&& array) {
+  static_assert(sizeof(To) == sizeof(From), "incompatible coercion");
+  Array<wchar_t> result;
+  memcpy(&result, &array, sizeof(array));
+  memset(&array, 0, sizeof(array));
```
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/encoding.c++]: reinterpreting objects as byte arrays -- as memcpy and memset do -- is actually blessed as defined behavior in the standard. I would expect a static analyzer to see the memcpy/memset and decide to let it be. But of course there's no real rules about what people can put in a static analyzer so I guess we'll see.

[the remark this answers is not in the document]
Comment (harrishancock) [reply c++/src/kj/encoding.c++]: I just ran clang-analyzer (`scan-build`) on this and it didn't complain, so I won't either. :)

For reference, my worry arose from this line from http://en.cppreference.com/w/cpp/string/byte/memcpy:

> If the objects are not TriviallyCopyable, the behavior of memcpy is not specified and may be undefined.

After researching the standard, that line seems unhelpfully imprecise.

[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/encoding.c++]: Eh, I think technically reinterpreting between `char16_t*` and `wchar_t*` is also undefined in itself (aliasing rules). But I'm not going to start making a copy of the whole buffer... so, oh well.

**Pull request merged**